test/induction.lean

```diff
@@ -0,0 +1,1338 @@
+import tactic.induction
+import tactic.linarith
+
+universes u v w
+
+inductive le : ℕ → ℕ → Type
+| zero {n} : le 0 n
+| succ {n m} : le n m → le (n + 1) (m + 1)
+
+inductive lt : ℕ → ℕ → Type
+| zero {n} : lt 0 (n + 1)
+| succ {n m} : lt n m → lt (n + 1) (m + 1)
+
+inductive Fin : ℕ → Type
+| zero {n} : Fin (n + 1)
+| succ {n} : Fin n → Fin (n + 1)
+
+inductive List (α : Sort*) : Sort*
+| nil {} : List
+| cons {} (x : α) (xs : List) : List
+
+namespace List
+
+def append {α} : List α → List α → List α
+| nil ys := ys
+| (cons x xs) ys := cons x (append xs ys)
+
+end List
+
+inductive Vec (α : Sort u) : ℕ → Sort (max 1 u)
+| nil : Vec 0
+| cons {n} : α → Vec n → Vec (n + 1)
+
+namespace Vec
+
+inductive eq {α} : ∀ n m, Vec α n → Vec α m → Prop
+| nil : eq 0 0 nil nil
+| cons {n m} {xs : Vec α n} {ys : Vec α m} {x y : α} :
+  x = y →
+  eq n m xs ys →
+  eq (n + 1) (m + 1) (cons x xs) (cons y ys)
+
+end Vec
+
+inductive Two : Type | zero | one
+
+inductive ℕ' : Type
+| intro : ℕ → ℕ'
+
+example (k) : 0 + k = k :=
+begin
+  induction' k,
+  { refl },
+  { simp }
+end
+
+example {k} (fk : Fin k) : Fin (k + 1) :=
+begin
+  induction' fk,
+  { exact Fin.zero },
+  { exact Fin.succ ih }
+end
+
+example {α} (l : List α) : l.append List.nil = l :=
+begin
+  induction' l,
+  { refl },
+  { dsimp only [List.append],
+    exact (congr_arg _ ih)
+  }
+end
+
+example {k l} (h : lt k l) : le k l :=
+begin
+  induction' h,
+  { exact le.zero },
+  { exact le.succ ih }
+end
+
+example {k l} : lt k l → le k l :=
+begin
+  induction' k; induction' l; intro hlt,
+  { cases' hlt },
+  { exact le.zero },
+  { cases' hlt },
+  { cases' hlt,
+    exact le.succ (@ih m hlt),
+  }
+end
+
+example {α n m} {xs : Vec α n} {ys : Vec α m} (h : Vec.eq n m xs ys) : n = m :=
+begin
+  induction' h,
+  case nil {
+    refl
+  },
+  case cons {
+    exact congr_arg nat.succ ih,
+  }
+end
+
+-- A simple induction with complex index arguments.
+example {k} (h : lt (k + 1) k) : false :=
```

Comment:
If an induction hypothesis has been simplified to `false`, should your tactic just close that goal? From a user perspective that case is probably not that different than if an index equation is simplified to false, which closes the goal (in e.g. the example of line 131)

Reply:
I'd find that mildly confusing. Right now, cases are closed automatically if and only if they are impossible due to the index values, which I think is a nice specification for what happens. You're probably right, though, that most users would also be happy with the goal closed for them. I don't have a strong opinion on this.

Reply:
Maybe you're right, and it is confusing. I'm happy with the status quo.